src/PermissionRegistrar.php

```diff
@@ -114,27 +111,59 @@ public function clearClassPermissions()
         $this->permissions = null;
     }
 
+    /**
+     * Load permissions from cache
+     * This get cache and turns array into \Illuminate\Database\Eloquent\Collection
+     */
+    private function loadPermissions()
+    {
+        if ($this->permissions === null) {
+            $this->permissions = $this->cache->remember(self::$cacheKey, self::$cacheExpirationTime, function () {
+                $permissions = $this->getPermissionClass()->select('id', 'name', 'guard_name')
+                    ->with('roles:id,name,guard_name')
+                    ->get()->toArray();
+                foreach ($permissions as $i => $permission) {
+                    foreach ($permission['roles'] ?? [] as $j => $roles) {
+                        unset($permissions[$i]['roles'][$j]['pivot']);
+                    }
+                }
+                return $permissions;
+            });
+            if (is_array($this->permissions)) {
+                $permissions = new Collection();
+                foreach ($this->permissions as $value) {
+                    $permissions->push($this->permissionClass::getModelFromArray($value));
```

Comment:
@erikn69, @drbyte 
Thanks for the improvements. Is there a reason why new methods were not added to the Role and Permission interface?

Reply:
Adding methods to an interface is a BC break, wich methods ? `getModelFromArray`?

Reply:
@erikn69 I'd accept a PR which updates the interface, since I'm leaning towards pushing the currently-open PRs into a v5.0 ... so your PR could be safely merged into that.

Reply:
Backward compatibility already broken since 4.3. if used own models for roles and permissions, a fatal error is thrown due to the lack of the getModelFromArray method. It looks like now it makes no sense to wait for the 5th version to change the interface.

Reply:
@drbyte How about I put a condition so that it only uses the cache trick only when extending from spatie models? Just for now

Reply:
Yes, that should solve the interface issue and fix the BC

Reply:
@drbyte fixed on #1824

Reply:
@incrize, do you mind testing the fix? It's now merged into `dev-master`.

Reply:
#1826 @drbyte maybe better and permanent solution

Reply:
Checked out changes #1826 locally, all my tests pass.